### PR TITLE
[stdlib] Set, Dictionary: Mix the hash table size into the hash seed

### PR DESCRIPTION
**Pull request by lorentey**
This is done by xoring the bucket count of the hash table to the hash seed. This only flips a single bit in the seed, but avalanche effects make that more than enough to break the correlation between element orderings in tables of different sizes.

This eliminates quadratic performance issues when we copy elements between hash tables of different sizes.

Fixes https://bugs.swift.org/browse/SR-3268

### CODE REVIEW

**Comment by lorentey** [comment]
I think this will need adjustments in non-verbatim bridging at minimum. Let's see if tests uncover anything else.

@swift-ci please test

**Comment by lorentey** [comment]
@swift-ci please smoke benchmark

**Comment by natecook1000** [comment]
Doesn't using the storage address break copy-on-write?

```swift
let a = [1: "one", 2: "two"]
var b = a
b[2] = "deux"        
```
The last line creates a new block of storage with a different base address, but just copies over the contents. Without rehashing and re-ordering everything, a different perturbation will mean misses.

**Comment by lorentey** [comment]
@natecook1000 Oh yeah, oops! COW copies must inherit the original's seed -- not even indices would work otherwise. Luckily, restricting the change to per-capacity seeding would work just fine, because capacity reallocations are allowed to break indices. 🎉

We can reenable per-"instance" seeding later, but we'll need to add the seed as a storage class ivar, and make sure to reuse the same seed when we're making a COW copy.

**Comment by lorentey** [comment]
@swift-ci please test

**Comment by lorentey** [comment]
@swift-ci please smoke benchmark

**Comment by lorentey** [comment]
@swift-ci please test

**Comment by lorentey** [comment]
@swift-ci please smoke benchmark

File: stdlib/public/core/HashedCollections.swift.gyb [hunk 1]
```diff
@@ -3989,10 +3989,18 @@ extension _Native${Self}Buffer
...
+    // We assign a unique seed to each distinct hash table size, so that we
+    // prevent the quadratic copy problem without breaking value semantics.
+    // See https://bugs.swift.org/browse/SR-3268
+    //
+    // We don't need to generate brand new seed; changing a few bits in the
+    // global seed by XORing the bucket count prevents the issue just as well.
+    //
+    // FIXME: Use true per-instance seeding. This requires adding the local seed
+    // into the storage class, because COW copies need to use the same seed as
+    // their original instance.
```
Comment (lorentey) [inline stdlib/public/core/HashedCollections.swift.gyb]: Wow, I can't speaks English good 😳 (I'll fix it when the tests are done.)

**Comment by lorentey** [comment]
This seems to be a performance win, and not just for the known quadratic cases (DictionaryFilter and -Copy). ✨

I did not expect to see those two regressions. They're not too bad, it's just weird that they exist.

File: validation-test/stdlib/Dictionary.swift [hunk 2]
```diff
@@ -4574,6 +4574,37 @@ DictionaryTestSuite.test("removeAt") {
...
+  // would be sorted into irregular slots in the smaller table.
+  let slice = large.prefix(2 * count / 100).dropFirst(count / 100)
+  var small = [Int: Int](minimumCapacity: large.capacity / 2)
+  expectLT(small.capacity, large.capacity)
+  for (key, value) in slice {
+    small[key] = value
+  }
+  // If this test fails, there is a problem with local hash seeding.
+  expectFalse(small.map{$0.key}.elementsEqual(slice.map{$0.key}))
+}
```
Comment (natecook1000) [inline validation-test/stdlib/Dictionary.swift]: This is a super cool test 👏👏

Comment (lorentey) [reply validation-test/stdlib/Dictionary.swift]: Collision chains at the 1% mark will still make this sometimes pass without correct seeding. ;(

Luckily, I think we can fix this by simply ignoring the first "few" copied elements.

**Comment by lorentey** [comment]
@swift-ci please test

**Comment by lorentey** [comment]
@swift-ci please benchmark

**Comment by lorentey** [comment]
@swift-ci please test linux platform

**Comment by lorentey** [comment]
@swift-ci benchmark

File: validation-test/stdlib/Dictionary.swift (hunk 2 above)
Comment (natecook1000) [reply validation-test/stdlib/Dictionary.swift]: I think we could also reduce the number of chains by allocating more for the big dictionary, too—if you start with `[Int: Int](minimumCapacity: count * 3)` or something there should be enough space that the collision chains are rare/short.

**Comment by lorentey** [comment]
@swift-ci please test

**Comment by lorentey** [comment]
@please benchmark

**Comment by lorentey** [comment]
@please benchmark

**Comment by lorentey** [comment]
@swift-ci please benchmark

**Comment by lorentey** [comment]
@swift-ci please benchmark

**Comment by lorentey** [comment]
@swift-ci test

**Comment by lorentey** [comment]
@swift-ci test

**Comment by lorentey** [comment]
@swift-ci benchmark

**Comment by lorentey** [comment]
@swift-ci benchmark

**Comment by lorentey** [comment]
Let's try again with https://github.com/apple/swift/pull/15300

@swift-ci benchmark

**Comment by lorentey** [comment]
Fifty-third time is the charm! 🍀

@swift-ci benchmark

**Comment by lorentey** [comment]
@swift-ci test

**Comment by lorentey** [comment]
@swift-ci test

**Comment by lorentey** [comment]
@swift-ci benchmark

**Comment by lorentey** [comment]
@swift-ci benchmark

**Comment by lorentey** [comment]
@swift-ci benchmark

**Comment by lorentey** [comment]
@swift-ci benchmark

**Comment by lorentey** [comment]
@swift-ci benchmark

**Comment by lorentey** [comment]
Wow, these benchmark results are noisy. `DictionaryFilter` and `DictionaryCopy` may be the only meaningful results here.

* `Sim2DArray`/`Hanoi`/`CharIteration*`/`CharIndexing*`/`RangeIterationSigned` don't use dictionaries
* `WordSplitASCII`/`WordSplitUTF16` is known to be flaky at the moment (#15337)
* `Histogram` seems to be sensitive to the random hash seed (I'll submit a PR to disable it while benchmarking for now; a full fix needs an update to `Dictionary`'s storage representation.)

LGTM! 🎉

**Pull request merged**